PS-7286: Add max allowed value check for encryption_key_id (8.0)

oleksandr-kachan opened the pull request:
There is no max allowed value check for encryption_key_id
while creating/altering table or tablespace. Because of this
the resulting value may contain any garbage in case provided value
exceeds underlying datatype limits.

Added check for max allowed value provided for encryption_key_id
in parser to be not bigger then UINT_MAX32-1. Added allowed values
check for innodb_default_encryption_key_id variable.

oleksandr-kachan commented:
https://ps80.cd.percona.com/job/percona-server-8.0-param/793/

satya-bodapati commented:
@oleksandr-kachan is this PR still valid after your changes? https://github.com/percona/percona-server/pull/3882/
If it is valid, please take over
If it is not valid, please bring those testcases to this PR

oleksandr-kachan commented:
> @oleksandr-kachan is this PR still valid after your changes? #3882
> If it is valid, please take over
> If it is not valid, please bring those testcases to this PR

@satya-bodapati 
Updated this PR, will close  #3882

oleksandr-kachan commented:
https://ps80.cd.percona.com/job/percona-server-8.0-param/794/

satya-bodapati commented:
> > @oleksandr-kachan is this PR still valid after your changes? #3882
> > If it is valid, please take over
> > If it is not valid, please bring those testcases to this PR
> 
> @satya-bodapati
> Updated this PR, will close #3882

Did you take any parts of code from #3882?

oleksandr-kachan commented:
> Did you take any parts of code from #3882?

@satya-bodapati 
Yes, I added check for min/max value for innodb_default_encryption_key_id variable but with my improvements. Also replaced my test with one from #3882.

satya-bodapati commented on mysql-test/suite/sys_vars/t/innodb_default_encryption_key_id_basic.test:
```diff
@@ -1,3 +1,62 @@
+CREATE TABLE t_MAX (a varchar(255)) ENCRYPTION_KEY_ID=4294967294;
+CREATE TABLE t_MIN (a varchar(255)) ENCRYPTION_KEY_ID=0;
+
+--error ER_ENCRYPTION_KEY_ID_VALUE_OUT_OF_RANGE
+ALTER TABLE t_MAX ENCRYPTION_KEY_ID=4294967295;
+--error ER_PARSE_ERROR
+ALTER TABLE t_MIN ENCRYPTION_KEY_ID=-1;
```
Why don't we get the same error code for both cases?

oleksandr-kachan replied:
@satya-bodapati, min and max value checks are done in different places in this case. For max value we directly check in parser rules which we can change and use whichever error we want. The min value check is done by parser itself based on datatype assigned to the symbol which is unsigned in this case - "ENCRYPTION_KEY_ID_SYM opt_equal real_ulong_num". We could in general change datatype for ENCRYPTION_KEY_ID_SYM to signed and do check its min value manually. Do you think we should follow this way?

satya-bodapati replied:
Got it! thanks

satya-bodapati commented on mysql-test/suite/sys_vars/r/innodb_default_encryption_key_id_basic.result:
```diff
@@ -1,3 +1,35 @@
 0
+# restart:--early-plugin-load=keyring_file=keyring_file.so --keyring_file_data=MYSQL_TMP_DIR/keyring/keyring --default-table-encryption=ONLINE_TO_KEYRING --innodb-encryption-rotate-key-age=15 --innodb-encryption-threads=4 --innodb-default-encryption-key-id=4294967295 --log-error=MYSQL_TMP_DIR/my_restart.err
+# restart:--early-plugin-load=keyring_file=keyring_file.so --keyring_file_data=MYSQL_TMP_DIR/keyring/keyring --default-table-encryption=ONLINE_TO_KEYRING --innodb-encryption-rotate-key-age=15 --innodb-encryption-threads=4 --innodb-default-encryption-key-id=-1 --log-error=MYSQL_TMP_DIR/my_restart.err
+SET innodb_default_encryption_key_id=4294967295;
+Warnings:
+Warning	1292	Truncated incorrect innodb_encryption_key_id value: '4294967295'
```
after every SET, please show the current value and global SELECT @@session. and @@global.

satya-bodapati commented on mysql-test/suite/sys_vars/t/innodb_default_encryption_key_id_basic.test:
```diff
@@ -1,3 +1,62 @@
+--let SEARCH_PATTERN=\[Server\] option 'innodb-default-encryption-key-id': unsigned value 4294967295 adjusted to 4294967294\.
+--source include/search_pattern_in_file.inc
+
+--let $restart_parameters=restart:--early-plugin-load="keyring_file=$KEYRING_PLUGIN" --keyring_file_data=$MYSQL_TMP_DIR/keyring/keyring --default-table-encryption=ONLINE_TO_KEYRING --innodb-encryption-rotate-key-age=15 --innodb-encryption-threads=4 --innodb-default-encryption-key-id=-1 --log-error=$error_log
```
where the keyring_file_data deleted?

satya-bodapati commented on mysql-test/suite/sys_vars/t/innodb_default_encryption_key_id_basic.test:
```diff
@@ -1,3 +1,62 @@
-
 SELECT @@global.innodb_default_encryption_key_id;
+
+--let $error_log=$MYSQL_TMP_DIR/my_restart.err
```
This should be also deleted? (or is it already deleted somewhere below?)

satya-bodapati commented on share/messages_to_clients.txt:
```diff
@@ -9480,6 +9480,9 @@ ER_SYSTEM_KEY_ROTATION_CANT_GENERATE_NEW_VERSION
 ER_DTE_ENCRYPTION_THREADS_ACTIVE
   eng "In order to change default_table_encryption, disable encryption threads. Please set number of innodb_encryption_threads to 0."
 
+ER_ENCRYPTION_KEY_ID_VALUE_OUT_OF_RANGE
+  eng "The requested value for the encryption_key_id exceeds the maximum allowed %lu."
```
```suggestion
ER_ENCRYPTION_KEY_ID_VALUE_OUT_OF_RANGE
  eng "The requested value %lu for the encryption_key_id exceeds the maximum allowed 4294967295"
```

oleksandr-kachan replied:
Done

satya-bodapati commented on mysql-test/suite/sys_vars/t/innodb_default_encryption_key_id_basic.test:
```diff
@@ -1,3 +1,108 @@
+SET global innodb_default_encryption_key_id=0;
+SELECT @@session.innodb_default_encryption_key_id;
+SELECT @@global.innodb_default_encryption_key_id;
+
```
can you set to non-zero before SET PERSIST? that way we will know whether SET PERSIST changed the value or not?

satya-bodapati replied:
Same for PERSIST_ONLY

satya-bodapati reviewed (changes requested):
Looks good mostly. One last minor request

satya-bodapati commented on mysql-test/suite/sys_vars/r/innodb_default_encryption_key_id_basic.result:
```diff
@@ -1,3 +1,140 @@
+SELECT @@global.innodb_default_encryption_key_id;
+@@global.innodb_default_encryption_key_id
+1
+SET persist innodb_default_encryption_key_id=0;
+SELECT @@session.innodb_default_encryption_key_id;
+@@session.innodb_default_encryption_key_id
+1
+SELECT @@global.innodb_default_encryption_key_id;
+@@global.innodb_default_encryption_key_id
+0
```
is this expected SET PERSIST behaviour? This changes the global value and not the session value. Can you please check SET PERSIST with another session variable?

oleksandr-kachan replied:
@satya-bodapati, to my knowledge this is expected - SET PERSIST updates global variable only.

satya-bodapati replied:
ok. Can you also check the Jenkins results. the test is skipped on release builds
201002 11:06:48 [ 77%] sys_vars.innodb_default_encryption_key_id_basic w4  [ skipped ]  Not implemented

Please resubmit to Jenkins (some issue with debug builds)

oleksandr-kachan commented:
https://ps80.cd.percona.com/job/percona-server-8.0-param/795/

satya-bodapati commented:
Analyzed Jenkins results. Looks good
